File: java/gov/nasa/gsfc/gmsec/api/mist/SchemaIDIterator.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file SchemaIDIterator.java
 */

package gov.nasa.gsfc.gmsec.api.mist;

import gov.nasa.gsfc.gmsec.api.GMSEC_Exception;

import gov.nasa.gsfc.gmsec.api.jni.mist.JNISchemaIDIterator;


/**
 * Supports the iteration over the list of schema templates maintained by a Specification object.
 *
 * @see Specification
 */
public class SchemaIDIterator
{
	private JNISchemaIDIterator m_jniSchemaIDIterator = null;


	private SchemaIDIterator()
	{
	}


	/** 
	 * This method is for internal GMSEC API use only.
	 * @param iter Object to reference for acquiring internal JNISchemaIDIterator.
	 * @return Internal JNISchemaIDIterator object.
	 */
	public static JNISchemaIDIterator getInternal(SchemaIDIterator iter)
	{
		return (iter == null ? null : iter.m_jniSchemaIDIterator);
	}


	/**
	 * This constructor is for internal GMSEC API use only.
	 * @param jIter Internal JNISchemaIDIterator object.
	 */
	public SchemaIDIterator(JNISchemaIDIterator jIter)
	{
		m_jniSchemaIDIterator = jIter;
	}


	/**
	 * Indicates whether there are additional Schema IDs that can be referenced using next().
	 *
	 * @return True is returned if additional fields are available, false otherwise.
	 */
	public boolean hasNext()
	{
		return m_jniSchemaIDIterator.hasNext();
	}


	/**
	 * Returns the next available Schema ID.
	 *
	 * @return A Schema ID string.
	 *
	 * @throws GMSEC_Exception Thrown if the iterator has reached the end (i.e. there are no more IDs).
	 */
	public String next() throws GMSEC_Exception
	{
		return m_jniSchemaIDIterator.next();
	}


	/**
	 * Resets the iterator to the beginning of the schema list that is maintained by the Specification object.
	 */
	public void reset()
	{
		m_jniSchemaIDIterator.reset();
	}
}

File: framework/src/gmsec4/c/mist/schema_id_iterator.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file schema_id_iterator.cpp
 *
 * @brief This file contains function for the management of a SchemaIDIterator object.
 */


#include <gmsec4/c/mist/schema_id_iterator.h>

#include <gmsec4/Exception.h>
#include <gmsec4/Status.h>

#include <gmsec4/mist/SchemaIDIterator.h>


using namespace gmsec::api;
using namespace gmsec::api::mist;


GMSEC_BOOL CALL_TYPE schemaIDIteratorHasNext(const GMSEC_SchemaIDIterator iter, GMSEC_Status status)
{
	const SchemaIDIterator* idIter = reinterpret_cast<const SchemaIDIterator*>(iter);
	GMSEC_BOOL              hasNext = GMSEC_FALSE;
	Status                  result;

	if (idIter)
	{
		hasNext = (idIter->hasNext() ? GMSEC_TRUE : GMSEC_FALSE);
	}
	else
	{
		result = Status(ITERATOR_ERROR, UNINITIALIZED_OBJECT, "SchemaIDIterator handle is NULL");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return hasNext;
}


const char* CALL_TYPE schemaIDIteratorNext(GMSEC_SchemaIDIterator iter, GMSEC_Status status)
{
	SchemaIDIterator* idIter   = reinterpret_cast<SchemaIDIterator*>(iter);
	const char*       schemaID = NULL;
	Status            result;

	if (idIter)
	{
		try
		{
			schemaID = idIter->next();
		}
		catch (Exception& e)
		{
			result = Status(e);
		}
	}
	else
	{
		result = Status(ITERATOR_ERROR, UNINITIALIZED_OBJECT, "SchemaIDIterator handle is NULL");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return schemaID;
}


void CALL_TYPE schemaIDIteratorReset(GMSEC_SchemaIDIterator iter, GMSEC_Status status)
{
	SchemaIDIterator* idIter = reinterpret_cast<SchemaIDIterator*>(iter);
	Status            result;

	if (idIter)
	{
		idIter->reset();
	}
	else
	{
		result = Status(ITERATOR_ERROR, UNINITIALIZED_OBJECT, "SchemaIDIterator handle is NULL");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}
}
